### PR TITLE
[#169546582] Update approach to load metadata of services to get the scope

### PR DESCRIPTION
**Pull request by valentinacala**
**Short description:**
this pr updates the approach used to load the scope of the visible services (used to display the proper content into the service home screen tabs).

**List of changes proposed in this pull request:**
- Update the startup saga so that the actions and sagas related to the  load/refresh of services are injected in the new `watchLoadServicesSaga`
- Update the loadVisibleServicesRequestHandler saga by introducing `removeOldStoredServices` and `refreshStoredServices` to make the code being more readable  
- Add tests related to loading of visible services (`loadVisibleServicesHandler.test.ts`) and service content (`loadServiceContentRequestHandler.test.ts`) + test of the new  fuctions used to wrap existing code
- **[CORE CHANGE]** All the services metadata was previously loaded at startup. Instead of it, at startup it is loaded only the `servicesByScope` object, while all the metadata related to a certain service are loaded when the user accesses the service detail (from service home screen selection or from the message CTA bar).To apply this change, the following files have been updated:
   - `ServicesHomeScreen.tsx `
   - `store/reducers/entities/services/index.ts` and the corresponding test file

- **[CORE CHANGE 2]** the logic to manage the "isFirstServiceLoadCompleted" state is updated. The same logic has been wrapped in dedicated functions (`loadServiceDetailRequestHandler.test.ts`, `handleServiceReadabilitySaga.test.ts`, `handleFirstVisibleServiceLoadSaga.test.ts `) and they have been updated the following files:
     - ` contentLoaders.ts`: it manages the service metadata load - the logic was here
     - `loadServiceDetailRequestHandler.ts `: it manages the service detail load - it hosts now the logic

**How to test:**
Once logged in for the first time, by navigating to the service home screen the loading screen is displayed for a shorter time (or not displayed at all)

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#169546582](https://www.pivotaltracker.com/story/show/169546582): Recuperare le informazioni dello scope dei servizi dal nuovo endpoint

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: ts/sagas/startup/__tests__/loadServiceContentRequestHandler.test.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      .isDone();
+  });
+
+  it("returns service content if the backand response is 200", () => {
+    testSaga(loadServiceContentRequestHandler, getService, mockedAction)
+      .next()
+      .call(getService, { service_id: mockedServiceId })
+      .next(right({ status: 200, value: mockedService }))
+      .put(loadServiceContent.success(mockedService))
+      .next();
```
Comment (Undermaken) [inline ts/sagas/startup/__tests__/loadServiceContentRequestHandler.test.ts]:
```suggestion
      .next() // update organization name
      .next()
      .isDone();
```
in this way you can skip the update organization name logic but you can check that the generator ends successfully

File: ts/sagas/startup/__tests__/loadVisibleServicesHandler.test.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        .isDone();
+    });
+
+    it("return an array of visibile services if backend response is 200", () => {
+      testSaga(loadVisibleServicesRequestHandler, getVisibleServices)
+        .next()
+        .call(getVisibleServices, {})
+        .next(right({ status: 200, value: { items: mockedVisibleServices } }))
+        .put(loadVisibleServices.success(mockedVisibleServices))
+        .next();
```
Comment (Undermaken) [inline ts/sagas/startup/__tests__/loadVisibleServicesHandler.test.ts]:
```suggestion
       .call(removeOldStoredServices) // removeOldStoredServices
        .call(refreshStoredServices) // refreshStoredServices
        .next()
        .isDone();
```
(you have to export these functions for the test to work)
your refactoring allows to test if some action are dispatched and (if you want) to test these generators too

File: ts/sagas/startup.ts [hunk 3]
```diff
@@ -292,19 +292,7 @@ function* initializeApplicationSaga(): IterableIterator<Effect> {
...
-  );
-
-  yield takeEvery(
-    getType(loadVisibleServices.request),
-    loadVisibleServicesRequestHandler,
-    backendClient.getVisibleServices
-  );
-  // Trigger the services content and metadata being loaded/refreshed.
-  yield put(loadVisibleServices.request());
+  yield fork(watchLoadServicesSaga, backendClient);
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/startup.ts]: added wihtin https://github.com/teamdigitale/io-app/pull/1503/commits/ec9e3740a5f69fac1995c83ffbb550d3cb83640b

File: ts/sagas/startup/loadServiceContentRequestHandler.ts [hunk 4]
```diff
@@ -24,9 +23,9 @@ import { isVisibleService } from "../../utils/services";
  * @param {string} id - The id of the service to load
  * @returns {IterableIterator<Effect | Either<Error, ServicePublic>>}
  */
-export function* loadServiceRequestHandler(
+export function* loadServiceContentRequestHandler(
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/startup/loadServiceContentRequestHandler.ts]: updated within https://github.com/teamdigitale/io-app/commit/ec9e3740a5f69fac1995c83ffbb550d3cb83640b

File: ts/sagas/startup/loadVisibleServicesHandler.ts [hunk 5]
```diff
@@ -19,14 +18,18 @@ type VisibleServiceVersionById = {
   [index: string]: number | undefined;
 };
 
+type ReturnedVisibleServicesType = ReadonlyArray<{
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/startup/loadVisibleServicesHandler.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/a2d0c0edbf7e08a0b17c8d98e308781e4e89f712

File: ts/sagas/startup/__tests__/loadServiceContentRequestHandler.test.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        loadServiceContent.failure({
+          service_id: mockedServiceId,
+          error: mockedGenericError
+        })
+      )
+      .next()
+      .isDone();
+  });
+
+  it("returns service content if the backand response is 200", () => {
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/startup/__tests__/loadServiceContentRequestHandler.test.ts]: fixed within https://github.com/teamdigitale/io-app/pull/1503/commits/7912b800a90ba28ee09f166c234e1299d99da33f

File: ts/store/actions/content.ts [hunk 7]
```diff
@@ -34,13 +33,13 @@ export const contentMunicipalityLoad = createAsyncAction(
   Error
 >();
 
-export const contentServicesByScopeLoad = createAsyncAction(
-  "CONTENT_SERVICES_BY_SCOPE_LOAD_REQUEST",
-  "CONTENT_SERVICES_BY_SCOPE_LOAD_SUCCESS",
-  "CONTENT_SERVICES_BY_SCOPE_LOAD_FAILURE"
+export const metadataServicesByScopeLoad = createAsyncAction(
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/store/actions/content.ts]: Being an invasive change, to make things simple I opted to remove the changes to the action naming and open a new story for the renaming of the actions.

File: ts/sagas/startup/__tests__/loadServiceContentRequestHandler.test.ts (hunk 1 above)
Comment (valentinacala) [reply ts/sagas/startup/__tests__/loadServiceContentRequestHandler.test.ts]: To improve the implementation, I moved the code about the organization name check into a dedicated function (https://github.com/teamdigitale/io-app/pull/1503/commits/a1f5fafde768f31ddf2c6f23126aa5ce0772268a), updating the existing `loadServiceContentRequestHandler.test.ts` and adding the test also for the new function.

File: ts/sagas/startup/__tests__/loadVisibleServicesHandler.test.ts (hunk 2 above)
Comment (valentinacala) [reply ts/sagas/startup/__tests__/loadVisibleServicesHandler.test.ts]: @Undermaken  I opted to create a dedicated store/services folder where store each function on a dedicated file, adding also the code for testing most of the isolated functions. I had problems only with testing the `removeOldStoredServices` function*. just trying with the most simple test such as 
```
import { testSaga } from "redux-saga-test-plan";
import { removeOldStoredServices } from '../removeOldStoredServices';

describe("removeOldStoredServices", () => {

    it("does something", () => {
        testSaga(removeOldStoredServices, {})
        .next()
    })
});
```
it returns the following error:

`TypeError: visibleServices.reduce is not a function`

It seems like the problem is with the reduce method if it is applied to `PaginatedServiceTupleCollection["items"]` objects.

We could proceed by adding the test of the mentioned function with a new pr, having to review the refresh/clean process. Sounds good?

Comment (Undermaken) [reply ts/sagas/startup/__tests__/loadVisibleServicesHandler.test.ts]: @valentinacala  Yes it does!
Track it in a new story please.
Thanks ;)

File: ts/sagas/services/_tests_/handleOrganizationNameUpdateSaga.test.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    testSaga(handleOrganizationNameUpdateSaga, mockedService)
+      .next()
+      .select(organizationNamesByFiscalCodeSelector)
+      .next(mockedOrganizationsNameByFiscalCode)
+      .select(visibleServicesSelector)
+      .next(mockedVisibleServices)
+      .put(updateOrganizations(mockedService))
+      .next()
+      .isDone();
+  });
```
Comment (Undermaken) [inline ts/sagas/services/_tests_/handleOrganizationNameUpdateSaga.test.ts]: Has it sense to add this test?

```typescript
it("does nothing if the organizationNamesByFiscalCodeSelector return undefined", () => {
    testSaga(handleOrganizationNameUpdateSaga, mockedService)
      .next()
      .select(organizationNamesByFiscalCodeSelector)
      .next(undefined)
      .isDone();
  });
```

File: ts/sagas/services/handleServiceReadabilitySaga.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import { ServiceId } from "../../../definitions/backend/ServiceId";
+import { markServiceAsRead } from "../../store/actions/services";
+import { isFirstVisibleServiceLoadCompletedSelector } from "../../store/reducers/entities/services/firstServicesLoading";
+
+/**
+ * A function to check if a service is loaded for the first time (at first startup or when the cache
+ * is cleaned). If true, the app shows the service list item without badge
+ * @param serviceId
+ */
+export function* handleServiceReadabilitySaga(serviceId: string): SagaIterator {
```
Comment (Undermaken) [inline ts/sagas/services/handleServiceReadabilitySaga.ts]:
```suggestion
export function* handleServiceReadabilitySaga(serviceId: ServiceId): SagaIterator {
```

File: ts/sagas/services/handleServiceReadabilitySaga.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  const isFirstVisibleServiceLoadCompleted: ReturnType<
+    typeof isFirstVisibleServiceLoadCompletedSelector
+  > = yield select(isFirstVisibleServiceLoadCompletedSelector);
+
+  if (!isFirstVisibleServiceLoadCompleted) {
+    yield put(markServiceAsRead(serviceId as ServiceId));
```
Comment (Undermaken) [inline ts/sagas/services/handleServiceReadabilitySaga.ts]:
```suggestion
    yield put(markServiceAsRead(serviceId));
```

File: ts/sagas/services/refreshStoredServices.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import * as pot from "italia-ts-commons/lib/pot";
+import { SagaIterator } from "redux-saga";
+import { all, put, select } from "redux-saga/effects";
+import { PaginatedServiceTupleCollection } from "../../../definitions/backend/PaginatedServiceTupleCollection";
+import { loadService } from "../../store/actions/services";
+import { servicesByIdSelector } from "../../store/reducers/entities/services/servicesById";
+
```
Comment (Undermaken) [inline ts/sagas/services/refreshStoredServices.ts]:
```typescript
/**
 * Check which services detail must be loaded. If there are, loading action will be dispatched
 * A service detail has to be loaded if one of these conditions is true
 * - it is not into the store (servicesByIdSelector)
 * - the relative stored value is a pot none (but not loading)
 * - the relative stored value is some and its version is less than the visible one
 * @param visibleServices
 */
```

File: ts/sagas/services/removeOldStoredServices.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
+import * as pot from "italia-ts-commons/lib/pot";
```
Comment (Undermaken) [inline ts/sagas/services/removeOldStoredServices.ts]: this file and function too, could be named **_removeUnusedStoredServices.ts_**
What do you think?

File: ts/store/reducers/entities/services/index.ts [hunk 13]
```diff
@@ -177,27 +163,27 @@ export const organizationsOfInterestSelector = createSelector(
...
-  if (pot.isSome(service)) {
-    const potServiceMetadata =
-      servicesMetadataById[service.value.service_id] || pot.none;
-    if (pot.isSome(potServiceMetadata) && potServiceMetadata.value) {
-      return potServiceMetadata.value.scope === scope;
-    }
+  // if service or servicesByScope are Error, the item is not included into the section
+  if (pot.isSome(service) && pot.isSome(servicesByScope)) {
+    const serviceId = service.value.service_id;
+    return scope === ScopeEnum.LOCAL
```
Comment (Undermaken) [inline ts/store/reducers/entities/services/index.ts]:
```suggestion
    return servicesByScope.value[scope].indexOf(serviceId) !== -1;
```

File: ts/sagas/services/handleServiceReadabilitySaga.ts (hunk 9 above)
Comment (valentinacala) [reply ts/sagas/services/handleServiceReadabilitySaga.ts]: I suggest to not update from `string` to `ServiceId` within this pr and postpone it for a new pr: the string type is inherited from the attribute `sender_service_id` of both `CreatedMessageWithoutContent` and `CreatedMessageWithContent` ( used to load services related to messages). What about update both the message-related types and, then, update also che app code?

File: ts/store/reducers/entities/services/index.ts (hunk 13 above)
Comment (valentinacala) [reply ts/store/reducers/entities/services/index.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/4081a008d43c026da44da6003142b34298c558bb

File: ts/sagas/services/removeOldStoredServices.ts (hunk 12 above)
Comment (valentinacala) [reply ts/sagas/services/removeOldStoredServices.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/4081a008d43c026da44da6003142b34298c558bb

File: ts/sagas/services/refreshStoredServices.ts (hunk 11 above)
Comment (valentinacala) [reply ts/sagas/services/refreshStoredServices.ts]: added within https://github.com/teamdigitale/io-app/pull/1503/commits/4081a008d43c026da44da6003142b34298c558bb

File: ts/sagas/services/handleServiceReadabilitySaga.ts (hunk 10 above)
Comment (valentinacala) [reply ts/sagas/services/handleServiceReadabilitySaga.ts]: related to https://github.com/teamdigitale/io-app/pull/1503#discussion_r369579954

File: ts/sagas/services/_tests_/handleOrganizationNameUpdateSaga.test.ts (hunk 8 above)
Comment (valentinacala) [reply ts/sagas/services/_tests_/handleOrganizationNameUpdateSaga.test.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/f52225941580107651d220e78aeb1506e6934fcb

File: ts/sagas/services/handleOrganizationNameUpdateSaga.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+): SagaIterator {
+  // If the organization fiscal code is associated to different organization names,
+  // it is considered valid the one declared for a visible service
+  const organizations: OrganizationNamesByFiscalCodeState = yield select(
+    organizationNamesByFiscalCodeSelector
+  );
+  if (organizations) {
+    const fc = service.organization_fiscal_code;
+    const organization = organizations[fc];
+    const visibleServices: VisibleServicesState = yield select(
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/services/handleOrganizationNameUpdateSaga.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/f52225941580107651d220e78aeb1506e6934fcb

File: ts/sagas/services/_tests_/handleFirstVisibleServiceLoadSaga.test.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
+import * as pot from "italia-ts-commons/lib/pot";
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/services/_tests_/handleFirstVisibleServiceLoadSaga.test.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/f52225941580107651d220e78aeb1506e6934fcb

File: ts/sagas/services/_tests_/handleOrganizationNameUpdateSaga.test.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
+import * as pot from "italia-ts-commons/lib/pot";
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/services/_tests_/handleOrganizationNameUpdateSaga.test.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/f52225941580107651d220e78aeb1506e6934fcb

File: ts/sagas/services/_tests_/handleFirstVisibleServiceLoadSaga.test.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+      .next()
+      .select(isFirstVisibleServiceLoadCompletedSelector)
+      .next(false)
+      .select(servicesByScopeSelector)
+      .next(mockedservicesByScope)
+      .select(visibleServicesDetailLoadStateSelector)
+      .next(pot.noneLoading)
+      .isDone();
+  });
+
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/services/_tests_/handleFirstVisibleServiceLoadSaga.test.ts]: added within https://github.com/teamdigitale/io-app/pull/1503/commits/f52225941580107651d220e78aeb1506e6934fcb

File: ts/sagas/services/handleFirstVisibleServiceLoadSaga.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+      typeof servicesByScopeSelector
+    > = yield select(servicesByScopeSelector);
+    const visibleServicesDetailsLoadState: ReturnType<
+      typeof visibleServicesDetailLoadStateSelector
+    > = yield select(visibleServicesDetailLoadStateSelector);
+    if (
+      servicesByScope &&
+      visibleServicesDetailsLoadState &&
+      pot.isSome(visibleServicesDetailsLoadState) &&
+      pot.isSome(servicesByScope)
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/sagas/services/handleFirstVisibleServiceLoadSaga.ts]: updated within https://github.com/teamdigitale/io-app/pull/1503/commits/f52225941580107651d220e78aeb1506e6934fcb. 

p.s. I found that using the `SagaIterator` type the false positive errors on the possible undefined returned object are resolved

File: ts/sagas/services/handleServiceReadabilitySaga.ts (hunk 9 above)
Comment (Undermaken) [reply ts/sagas/services/handleServiceReadabilitySaga.ts]: @valentinacala 
We will change it in a new PR

**Pull request merged**